dependency/vault_pki.go

```diff
@@ -66,40 +64,68 @@ func (d *VaultPKIQuery) Fetch(clients *ClientSet, opts *QueryOptions,
 	default:
 	}
 
-	var encPEM []byte
-	var err error
-	if encPEM, err = os.ReadFile(d.filePath); err != nil || len(encPEM) == 0 {
-		// no need to write to file as it is the template dest
-		encPEM, err = d.fetchPEM(clients)
-	}
-	if err != nil {
-		return nil, nil, errors.Wrap(err, d.String())
-	}
+	needsRenewal := fmt.Errorf("needs renewal")
+	getPEM := func(renew bool) ([]byte, error) {
+		var encPEM []byte
+		var err error
+		encPEM, err = os.ReadFile(d.filePath)
+		if renew || err != nil || len(encPEM) == 0 {
+			encPEM, err = d.fetchPEM(clients)
+			// no need to write cert to file as it is the template dest
+		}
+		if err != nil {
+			return nil, err
+		}
 
-	block, cert, err := getCert(encPEM)
-	if err != nil {
-		return nil, nil, err
+		block, cert, err := getCert(encPEM)
+		if err != nil {
+			return nil, err
+		}
+
+		if sleepFor, ok := goodFor(cert); ok {
+			d.sleepCh <- sleepFor
+			return pem.EncodeToMemory(block), nil
+		}
+		return []byte{}, needsRenewal
 	}
 
-	if sleepFor, ok := goodFor(cert); ok {
-		d.sleepCh <- sleepFor
-		return respWithMetadata(string(pem.EncodeToMemory(block)))
+	pemBytes, err := getPEM(false)
+	switch err {
+	case nil:
+	case needsRenewal:
+		pemBytes, err = getPEM(true)
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		return nil, nil, err
 	}
-	d.pemBlock = nil
-	return "", &ResponseMetadata{
-		LastContact: 0,
-		LastIndex:   0,
-	}, nil
+	return respWithMetadata(string(pemBytes))
 }
 
-// returns time left in 90% of the original lease
-// boolean returns false if time has already past
+// returns time left in ~90% of the original lease and a boolean
+// that returns false if cert needs renewing, true otherwise
 func goodFor(cert *x509.Certificate) (time.Duration, bool) {
 	start, end := cert.NotBefore.Unix(), cert.NotAfter.Unix()
-	// use 90% of the cert duration
-	goodfor := ((end - start) * 9) / 10
-	sleepFor := time.Until(time.Unix(start+goodfor, 0))
-	return sleepFor, sleepFor > 0
+	now := time.Now().UTC().Unix()
+	if end <= now || start >= now {
+		return 0, false // expired or not good yet
+	}
+	lifespan := end - start
+	duration := end - now          // duration remaining
+	gooddur := (duration * 9) / 10 // 90% of duration
+	mindur := (lifespan / 10)      // 10% of lifespan
+	if gooddur <= mindur {
+		return 0, false // almost expired, get a new one
+	}
+	if gooddur > 100 { // add jitter if big enough for it to matter
+		r := rand.New(rand.NewSource(time.Now().UnixNano()))
+		// between 87% and 92%
+		gooddur = gooddur + ((gooddur / 100) * int64(r.Intn(5)-3))
+	}
+	// with jitter,
+	sleepFor := time.Until(time.Unix(now+gooddur, 0))
```

Comment:
I could be wrong, but this looks equivalent to `sleepFor := gooddur`

Reply:
gooddur (int64) is the duration in seconds between now and NotAfter plus (at this point) the jitter. This adds back on the now to make the value the correct future time (as measured in seconds from epoch) and the time.Until call converts that into a time.Duration type which is what we want to return.

Reply:
Ah, I forgot about the duration. I think that would be equivalent to `gooddur * time.Second` then. But, either way, it works. :)

Reply:
Oh yeah.. didn't think of that. Good idea! Going to try it out for size and see if I like it better. Seems like it would simplify things and make them more obvious to the reader. Thanks!

Reply:
Oh... and do you think changing 'duration' to 'remaining' would make that any clearer?

BTW... Thanks for the review!

Reply:
Updated to create the time.Duration directly from the int64 seconds value.